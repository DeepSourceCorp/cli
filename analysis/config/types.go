package config

import "github.com/deepsourcelabs/cli/analysis/lsp"

///////////////////////////////////////////////////////
// DSConfig is the struct for .deepsource.toml file //
/////////////////////////////////////////////////////

type DSConfig struct {
	Version         int           `mapstructure:"version,omitempty" json:"version" toml:"version"`
	ExcludePatterns []string      `mapstructure:"exclude_patterns,omitempty" json:"exclude_patterns,omitempty" toml:"exclude_patterns"`
	TestPatterns    []string      `mapstructure:"test_patterns,omitempty" json:"test_patterns,omitempty" toml:"test_patterns"`
	Analyzers       []Analyzer    `mapstructure:"analyzers,omitempty" json:"analyzers,omitempty" toml:"analyzers"`
	Transformers    []Transformer `mapstructure:"transformers,omitempty" json:"transformers,omitempty" toml:"transformers"`
}

type Analyzer struct {
	Name                string      `mapstructure:"name,omitempty" json:"name,omitempty" toml:"name"`
	RuntimeVersion      string      `mapstructure:"runtime_version,omitempty" json:"runtime_version,omitempty" toml:"runtime_version,omitempty"`
	Enabled             bool        `mapstructure:"enabled,omitempty" json:"enabled" toml:"enabled"`
	DependencyFilePaths []string    `mapstructure:"dependency_file_paths,omitempty" json:"dependency_file_paths,omitempty"`
	Meta                interface{} `mapstructure:"meta,omitempty" json:"meta,omitempty" toml:"meta"`
	Thresholds          interface{} `mapstructure:"thresholds,omitempty" json:"thresholds,omitempty"`
}

type Transformer struct {
	Name    string `mapstructure:"name,omitempty" json:"name,omitempty" toml:"name"`
	Enabled bool   `mapstructure:"enabled,omitempty" json:"enabled,omitempty" toml:"enabled"`
}

///////////////////////////////////////
// LSP based Analysis Config Types  //
/////////////////////////////////////

type AnalysisConfig struct {
	Files           []lsp.TextDocumentItem `json:"files"`
	TestFiles       []lsp.TextDocumentItem `json:"test_files"`
	ExcludedFiles   []lsp.TextDocumentItem `json:"excluded_files"`
	ExcludePatterns []string               `json:"exclude_patterns"`
	TestPatterns    []string               `json:"test_patterns"`
	AnalyzerMeta    interface{}            `json:"analyzer_meta"`
}

//////////////////////////////////////
// LSP based Analysis Result Types  //
/////////////////////////////////////

type Namespace struct {
	Key   string  `json:"key"`
	Value float64 `json:"value"`
}

type Metric struct {
	MetricCode string      `json:"metric_code"`
	Namespaces []Namespace `json:"namespaces"`
}

type AnalysisResult struct {
	Issues   []lsp.Diagnostic `json:"issues"`
	Metrics  []Metric         `json:"metrics,omitempty"`
	IsPassed bool             `json:"is_passed"`
	Errors   []Error          `json:"errors"`
	// Errors    []lsp.Diagnostic `json:"errors"`
	ExtraData interface{} `json:"extra_data"`
}

////////////////////////////////////
// Default Analysis Result Types  //
///////////////////////////////////

type Error struct {
	HMessage string `json:"hmessage"`
	Level    int    `json:"level"`
}

type Coordinate struct {
	Line   int `json:"line"`
	Column int `json:"column"`
}

type Position struct {
	Begin Coordinate `json:"begin"`
	End   Coordinate `json:"end"`
}

type Location struct {
	Path     string   `json:"path"`
	Position Position `json:"position"`
}

type Issue struct {
	Code     string   `json:"issue_code"`
	Title    string   `json:"issue_text"`
	Location Location `json:"location"`
}

type DefaultAnalysisResult struct {
	Issues    []Issue     `json:"issues"`
	Metrics   []Metric    `json:"metrics,omitempty"`
	IsPassed  bool        `json:"is_passed"`
	Errors    []Error     `json:"errors"`
	ExtraData interface{} `json:"extra_data"`
}
