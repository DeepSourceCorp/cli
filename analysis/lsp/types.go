package lsp

//////////////////////////////
// Document identity types  //
/////////////////////////////

type DocumentURI string

type TextDocumentItem struct {
	URI        DocumentURI `json:"uri"`
	LanguageID string      `json:"languageID,omitempty"`
	Version    int         `json:"version,omitempty"`
	Text       string      `json:"text,omitempty"`
}

///////////////////////////////
// Diagnostic related types //
/////////////////////////////

type DiagnosticSeverity int

const (
	Error       DiagnosticSeverity = 1
	Warning     DiagnosticSeverity = 2
	Information DiagnosticSeverity = 3
	Hint        DiagnosticSeverity = 4
)

type Position struct {
	Line      int `json:"line"`
	Character int `json:"character"`
}
type Range struct {
	Start Position `json:"start"`
	End   Position `json:"end"`
}

type Location struct {
	URI   string `json:"uri"`
	Range Range  `json:"range"`
}

type DiagnosticRelatedInformation struct {
	Location Location `json:"location"`
	Message  string   `json:"message"`
}

type Diagnostic struct {
	Range    Range              `json:"range"`
	Severity DiagnosticSeverity `json:"severity,omitempty"`
	Code     string             `json:"code,omitempty"`
	Source   string             `json:"source,omitempty"`
	Message  string             `json:"message"`

	/**
	 * An array of related diagnostic information, e.g. when symbol-names within
	 * a scope collide all definitions can be marked via this property.
	 * var a,b
	 * a := 2
	 * Issues in line 1 and 2 are related.
	 */
	RelatedInformation []DiagnosticRelatedInformation `json:"relatedInformation"`
}
