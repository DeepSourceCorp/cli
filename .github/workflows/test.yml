name: Unit Tests

on:
  push:
    branches: [develop]
  pull_request:
    branches: [develop]

jobs:
  test:
    runs-on: ubuntu-latest
    env:
      NODE_AUTH_TOKEN: ${{secrets.GITHUB_TOKEN}}
      ZEAL_SECRET: ${{secrets.GITHUB_TOKEN}}

    steps:
      - name: Checkout Bifrost
        uses: actions/checkout@v2

      - name: Setting Up Node
        uses: actions/setup-node@v2
        with:
          node-version: '14.x'
          always-auth: true
          registry-url: https://npm.pkg.github.com/
          scope: '@deepsourcelabs'
          cache: 'yarn'

      - name: Install packages
        run: yarn

      - name: Run Test
        run: yarn test --coverage=true --coverageReporters=cobertura

      - name: Install DeepSource CLI
        run: curl https://deepsource.io/cli | sh

      - name: Set DEEPSOURCE_DSN
        run: |
          ./bin/deepsource report --analyzer test-coverage --key javascript --value-file ./coverage/cobertura-coverage.xml
        env:
          DEEPSOURCE_DSN: ${{ secrets.BIFROST_DSN }}
