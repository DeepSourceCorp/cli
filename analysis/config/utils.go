package config

import (
	"log"
	"os"
	"path"
	"path/filepath"
	"strings"

	"github.com/bmatcuk/doublestar/v4"
)

/* Walks the `CODE_PATH` directory and returns all the files other than the ones present
 * in the .git folder for analysis in the form of a string array */
func readAllFiles(codePath string) ([]string, error) {
	fileCount := 0

	allFiles := make([]string, 0)

	err := filepath.Walk(codePath,
		func(path string, fileInfo os.FileInfo, err error) error {
			if err != nil {
				return err
			}
			fileCount++

			/* Check the following before appending to the list of files:
			 * - Should not be a directory
			 * - The walked file should not be present in .git folder
			 * TODO: Check why fileCount != 1 condition is used here */
			if !fileInfo.IsDir() || !strings.HasPrefix(path, filepath.Join(codePath, ".git")) && fileCount != 1 {
				allFiles = append(allFiles, path)
			}
			return nil
		})
	if err != nil {
		return allFiles, err
	}
	return allFiles, nil
}

// Returns the slice of files matching certain glob patterns
func (r *AnalysisRun) getMatchingFiles(patterns []string) ([]string, error) {
	matchedFiles := make([]string, 0)

	// Return all the files if no exclude_patterns are configured
	if len(patterns) == 0 {
		return matchedFiles, nil
	}

	for _, file := range r.AnalysisFiles {
		for i := range patterns {
			match, err := doublestar.Match(path.Join(r.CodePath, patterns[i]), file)
			if err != nil {
				log.Printf("Failed to match the file %s with pattern %s\n", file, patterns[i])
			}
			if match {
				matchedFiles = append(matchedFiles, file)
				break
			}
		}
	}
	return matchedFiles, nil
}

// Filters the analysis files and removes the files matching the exclude_patterns from them
// TODO: Improve the logic here
func (r *AnalysisRun) filterAnalysisFiles() {
	filteredFiles := []string{}
	for _, file := range r.AnalysisFiles {
		for _, excludedFile := range r.ExcludedFiles {
			if file == excludedFile {
				break
			}
			filteredFiles = append(filteredFiles, file)
		}
	}
	r.AnalysisFiles = filteredFiles
}
