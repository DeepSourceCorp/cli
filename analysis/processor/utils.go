package processor

import (
	"sort"

	"github.com/deepsourcelabs/cli/types"
)

// sortIssuesByFile sorts the issues in an alphabetical order according to the filenames
// where they got reported.
func (p *ProcessAnalysisResults) sortIssuesByFile() {
	sort.Slice(p.AnalysisResult.Issues, func(i, j int) bool {
		return p.AnalysisResult.Issues[i].Location.Path < p.AnalysisResult.Issues[j].Location.Path
	})
}

// Prepare a map with unique filenames as key and the issue range for each file as value
// This is done to ensure fewer loops when processing issues.
type IssueRange struct {
	Filename   string // The file which has the issues
	BeginIndex int    // Array index in report.Issues where the particular issue starts
	EndIndex   int    // Array index in report.Issues where the particular issue ends
}

// GenerateIssueRangeSlice generates an array containing the issue ranges with respect to files
// that helps us to go through them and map them to the files where they got reported instead
// of opening the file for each of them. The generated index looks like this:

// [{analyzer.go 0 0} {autofix_patch.go 1 1} {difftool.go 2 2} {patch.patch 3 5} {proc_skip_cq_test.go 6 6}]

// Here, the first field if filename and the second and third fields are the index range in which the issues reported
// in these files lie in the sorted AnalyzerReport slice.
func createIssueFileRange(report types.AnalysisResult) []IssueRange {
	fileCount := 0 // for 1 file, 0 based indexing
	issuesRange := []IssueRange{}
	prevFilename := report.Issues[0].Location.Path

	issueRange := IssueRange{
		BeginIndex: 0,
		EndIndex:   len(report.Issues) - 1,
		Filename:   prevFilename,
	}
	issuesRange = append(issuesRange, issueRange)

	// Iterating over the issues and creating an array containing issues with index data about
	// the files in which those issues are present
	for i := 1; i < len(report.Issues); i++ {
		issue := report.Issues[i]
		currentFilename := issue.Location.Path

		// TODO: Check when this condition is implied
		if issue.Location.Position.End.Line == -1 {
			issue.Location.Position.End.Line = issue.Location.Position.Begin.Line
		}

		if currentFilename != prevFilename {
			fileCount++

			issueRange = issuesRange[fileCount-1]
			issueRange.EndIndex = i - 1
			issuesRange[fileCount-1] = issueRange

			// Create for the new file
			issueRange = IssueRange{
				Filename:   currentFilename,
				BeginIndex: i,
				EndIndex:   len(report.Issues) - 1,
			}

			issuesRange = append(issuesRange, issueRange)
			prevFilename = currentFilename
		}
	}
	return issuesRange
}

// formatLSPResultsToDefault converts the LSP based analysis results into the default format supported by DeepSource.
func (p *ProcessAnalysisResults) formatLSPResultsToDefault() {
	p.AnalysisResult.IsPassed = p.Report.IsPassed
	p.AnalysisResult.Metrics = append(p.Report.Metrics, p.AnalysisResult.Metrics...)
	p.AnalysisResult.Errors = append(p.Report.Errors, p.AnalysisResult.Errors...)

	// Appending the issues to the default format of Analysis report
	for _, issue := range p.Report.Issues {
		analysisIssue := types.Issue{
			IssueCode: issue.Code,
			IssueText: issue.Message,
			Location: types.Location{
				Path: issue.RelatedInformation[0].Location.URI,
				Position: types.Position{
					Begin: types.Coordinate{
						Line:   issue.Range.Start.Line,
						Column: issue.Range.Start.Character,
					},
					End: types.Coordinate{
						Line:   issue.Range.End.Line,
						Column: issue.Range.End.Character,
					},
				},
			},
		}
		p.AnalysisResult.Issues = append(p.AnalysisResult.Issues, analysisIssue)
	}
}
