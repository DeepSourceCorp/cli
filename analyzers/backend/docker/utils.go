package docker

import (
	"bufio"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"strings"
)

type ErrorLine struct {
	Error       string      `json:"error"`
	ErrorDetail ErrorDetail `json:"errorDetail"`
}

type ErrorDetail struct {
	Message string `json:"message"`
}

type DockerBuildResponse struct {
	Stream string `json:"stream"`
}

func CheckBuildResponse(rd io.Reader, showAllLogs bool) error {
	var lastLine []byte
	count := 0
	var currentStream string

	scanner := bufio.NewScanner(rd)
	for scanner.Scan() {
		lastLine = scanner.Bytes()
		d := &DockerBuildResponse{}
		err := json.Unmarshal(lastLine, d)
		if err != nil {
			return err
		}
		if d.Stream == "" || d.Stream == "\n" || strings.Contains(d.Stream, "--->") || strings.TrimSuffix(d.Stream, "\n") == currentStream {
			continue
		}
		currentStream = strings.TrimSuffix(d.Stream, "\n")
		if showAllLogs {
			fmt.Println(currentStream)
		}
		count++
	}

	errLine := &ErrorLine{}
	json.Unmarshal([]byte(lastLine), errLine)
	if errLine.Error != "" {
		return errors.New(errLine.Error)
	}
	return scanner.Err()
}


