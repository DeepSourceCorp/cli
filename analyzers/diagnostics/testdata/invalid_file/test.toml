name = "2do-Checker"
