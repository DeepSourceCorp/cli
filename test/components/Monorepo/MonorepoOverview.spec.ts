import { render } from '@testing-library/vue'

import MonorepoOverview from '~/components/Monorepo/MonorepoOverview.vue'

import { focusDirective } from '~/plugins/helpers/directives.client'

import { cartesian, generateBooleanProps } from '~/test/utils'
import { RepositoryKindChoices, VcsProviderChoices } from '~/types/types'

describe('[[ MonorepoOverview ]]', () => {
  const baseProps = {
    currentPage: 1,
    totalPageCount: 10,
    searchQuery: '',
    showAddSubRepositoryCta: false,
    subRepositoryList: Array.from({ length: 10 }).map((_, idx) => ({
      id: `test-id-${idx + 1}`,
      name: `sub-repository-${idx + 1}`,
      vcsProvider: VcsProviderChoices.Github,
      isActivated: idx > 5,
      isPrivate: idx < 5,
      latestCommitOid: `test-commit-oid-${idx + 1}`,
      defaultBranchName: idx > 5 ? 'master' : 'main',
      lastAnalyzedAt: '2022-11-20T00:00:00+05:30',
      kind: RepositoryKindChoices.Subrepo
    })),
    ownerLogin: 'deepsourcelabs'
  }

  const stubs = {
    LazyEmptyState: true,
    RepoCard: true,
    ZButton: true,
    ZIcon: true,
    ZInput: true,
    ZPagination: true
  }

  test('renders different states based on the `loading` and `showAddSubRepositoryCta` prop combinations', () => {
    const loadingOptions = generateBooleanProps('loading', false)
    const showAddSubRepositoryCtaOptions = generateBooleanProps('showAddSubRepositoryCta', false)

    cartesian(loadingOptions, showAddSubRepositoryCtaOptions).forEach((propCombination) => {
      const propsData = {
        ...baseProps,
        ...propCombination
      }

      const { html } = render(MonorepoOverview, {
        propsData,
        stubs
      })

      expect(html()).toMatchSnapshot(JSON.stringify(propCombination))
    })
  })

  describe('Empty states', () => {
    test('search query that does not yield a result', () => {
      const propsData = {
        ...baseProps,
        searchQuery: 'test',
        subRepositoryList: []
      }

      const { html } = render(MonorepoOverview, {
        propsData,
        stubs
      })

      expect(html()).toMatchSnapshot()
    })

    test('no activated sub-repositories', () => {
      const propsData = {
        ...baseProps,
        subRepositoryList: []
      }

      const { html } = render(
        MonorepoOverview,
        {
          propsData,
          stubs
        },
        (vue) => {
          vue.directive('focus', focusDirective)
        }
      )

      expect(html()).toMatchSnapshot()
    })
  })
})
