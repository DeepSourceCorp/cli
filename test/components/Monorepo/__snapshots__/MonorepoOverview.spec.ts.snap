// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[[ MonorepoOverview ]] Empty states no activated sub-repositories 1`] = `
"<div class=\\"space-y-4\\">
  <div class=\\"monorepo-empty-state-card-shadow mx-auto mt-36 max-w-md space-y-6 rounded-md border border-ink-200 bg-ink-300 p-8 lg:mt-52 lg:max-w-lg\\">
    <source srcset=\\"~/assets/images/ui-states/no-activated-sub-repositories-state.webp\\" type=\\"image/webp\\"> <img src=\\"~/assets/images/ui-states/no-activated-sub-repositories-state.png\\" alt=\\"No activated sub-repositories\\">
    <div class=\\"space-y-2\\">
      <h2 class=\\"text-base font-medium text-vanilla-100\\">Define a new sub-repository</h2>
      <p class=\\"text-sm font-normal text-vanilla-400\\">
        This monorepo has no activated sub-repositories yet. Define a new sub-repository to start
        analyzing your code.
      </p>
    </div>
    <!---->
  </div>
  <zpagination-stub page=\\"1\\" totalpages=\\"10\\" totalvisible=\\"5\\" hideforsinglepage=\\"true\\" class=\\"flex justify-center\\"></zpagination-stub>
</div>"
`;

exports[`[[ MonorepoOverview ]] Empty states search query that does not yield a result 1`] = `
"<div class=\\"space-y-4 max-w-3xl\\">
  <div class=\\"flex gap-x-2\\">
    <zinput-stub name=\\"\\" value=\\"test\\" label=\\"\\" type=\\"text\\" placeholder=\\"Search for a sub-repository\\" debouncedelay=\\"350\\" size=\\"small\\" backgroundcolor=\\"ink-300\\" errorbordercolor=\\"cherry\\" class=\\"w-full\\"></zinput-stub>
    <!---->
  </div>
  <lazyemptystate-stub png-image-path=\\"test-image-dir/test-image-file.png\\" title=\\"No results found for 'test'\\" webp-image-path=\\"test-image-dir/test-image-file.png\\" subtitle=\\"Please try changing your search query.\\"></lazyemptystate-stub>
  <zpagination-stub page=\\"1\\" totalpages=\\"10\\" totalvisible=\\"5\\" hideforsinglepage=\\"true\\" class=\\"flex justify-center\\"></zpagination-stub>
</div>"
`;

exports[`[[ MonorepoOverview ]] renders different states based on the \`loading\` and \`showAddSubRepositoryCta\` prop combinations: {"showAddSubRepositoryCta":false,"loading":false} 1`] = `
"<div class=\\"space-y-4 max-w-3xl\\">
  <div class=\\"flex gap-x-2\\">
    <zinput-stub name=\\"\\" value=\\"\\" label=\\"\\" type=\\"text\\" placeholder=\\"Search for a sub-repository\\" debouncedelay=\\"350\\" size=\\"small\\" backgroundcolor=\\"ink-300\\" errorbordercolor=\\"cherry\\" class=\\"w-full\\"></zinput-stub>
    <!---->
  </div>
  <div class=\\"space-y-3\\">
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-1\\" show-inactive-tag=\\"true\\" size=\\"small\\" id=\\"test-id-1\\" name=\\"sub-repository-1\\" vcsprovider=\\"GITHUB\\" isprivate=\\"true\\" latestcommitoid=\\"test-commit-oid-1\\" defaultbranchname=\\"main\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-2\\" show-inactive-tag=\\"true\\" size=\\"small\\" id=\\"test-id-2\\" name=\\"sub-repository-2\\" vcsprovider=\\"GITHUB\\" isprivate=\\"true\\" latestcommitoid=\\"test-commit-oid-2\\" defaultbranchname=\\"main\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-3\\" show-inactive-tag=\\"true\\" size=\\"small\\" id=\\"test-id-3\\" name=\\"sub-repository-3\\" vcsprovider=\\"GITHUB\\" isprivate=\\"true\\" latestcommitoid=\\"test-commit-oid-3\\" defaultbranchname=\\"main\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-4\\" show-inactive-tag=\\"true\\" size=\\"small\\" id=\\"test-id-4\\" name=\\"sub-repository-4\\" vcsprovider=\\"GITHUB\\" isprivate=\\"true\\" latestcommitoid=\\"test-commit-oid-4\\" defaultbranchname=\\"main\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-5\\" show-inactive-tag=\\"true\\" size=\\"small\\" id=\\"test-id-5\\" name=\\"sub-repository-5\\" vcsprovider=\\"GITHUB\\" isprivate=\\"true\\" latestcommitoid=\\"test-commit-oid-5\\" defaultbranchname=\\"main\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-6\\" show-inactive-tag=\\"true\\" size=\\"small\\" id=\\"test-id-6\\" name=\\"sub-repository-6\\" vcsprovider=\\"GITHUB\\" latestcommitoid=\\"test-commit-oid-6\\" defaultbranchname=\\"main\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-7\\" size=\\"small\\" id=\\"test-id-7\\" name=\\"sub-repository-7\\" vcsprovider=\\"GITHUB\\" isactivated=\\"true\\" latestcommitoid=\\"test-commit-oid-7\\" defaultbranchname=\\"master\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-8\\" size=\\"small\\" id=\\"test-id-8\\" name=\\"sub-repository-8\\" vcsprovider=\\"GITHUB\\" isactivated=\\"true\\" latestcommitoid=\\"test-commit-oid-8\\" defaultbranchname=\\"master\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-9\\" size=\\"small\\" id=\\"test-id-9\\" name=\\"sub-repository-9\\" vcsprovider=\\"GITHUB\\" isactivated=\\"true\\" latestcommitoid=\\"test-commit-oid-9\\" defaultbranchname=\\"master\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-10\\" size=\\"small\\" id=\\"test-id-10\\" name=\\"sub-repository-10\\" vcsprovider=\\"GITHUB\\" isactivated=\\"true\\" latestcommitoid=\\"test-commit-oid-10\\" defaultbranchname=\\"master\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
  </div>
  <zpagination-stub page=\\"1\\" totalpages=\\"10\\" totalvisible=\\"5\\" hideforsinglepage=\\"true\\" class=\\"flex justify-center\\"></zpagination-stub>
</div>"
`;

exports[`[[ MonorepoOverview ]] renders different states based on the \`loading\` and \`showAddSubRepositoryCta\` prop combinations: {"showAddSubRepositoryCta":false,"loading":true} 1`] = `
"<div class=\\"space-y-4 max-w-3xl\\">
  <div class=\\"h-8 animate-pulse bg-ink-300\\"></div>
  <div class=\\"space-y-3\\">
    <div class=\\"h-18 animate-pulse bg-ink-300\\"></div>
    <div class=\\"h-18 animate-pulse bg-ink-300\\"></div>
    <div class=\\"h-18 animate-pulse bg-ink-300\\"></div>
    <div class=\\"h-18 animate-pulse bg-ink-300\\"></div>
  </div>
  <zpagination-stub page=\\"1\\" totalpages=\\"10\\" totalvisible=\\"5\\" hideforsinglepage=\\"true\\" class=\\"flex justify-center\\"></zpagination-stub>
</div>"
`;

exports[`[[ MonorepoOverview ]] renders different states based on the \`loading\` and \`showAddSubRepositoryCta\` prop combinations: {"showAddSubRepositoryCta":true,"loading":false} 1`] = `
"<div class=\\"space-y-4 max-w-3xl\\">
  <div class=\\"flex gap-x-2\\">
    <zinput-stub name=\\"\\" value=\\"\\" label=\\"\\" type=\\"text\\" placeholder=\\"Search for a sub-repository\\" debouncedelay=\\"350\\" size=\\"small\\" backgroundcolor=\\"ink-300\\" errorbordercolor=\\"cherry\\" class=\\"w-full\\"></zinput-stub>
    <zbutton-stub color=\\"\\" iconcolor=\\"\\" label=\\"\\" loadinglabel=\\"\\" as=\\"button\\" justify=\\"center\\" buttontype=\\"primary\\" type=\\"button\\" size=\\"small\\" to=\\"\\" icon=\\"folder-plus\\">New sub-repository</zbutton-stub>
  </div>
  <div class=\\"space-y-3\\">
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-1\\" show-inactive-tag=\\"true\\" size=\\"small\\" id=\\"test-id-1\\" name=\\"sub-repository-1\\" vcsprovider=\\"GITHUB\\" isprivate=\\"true\\" latestcommitoid=\\"test-commit-oid-1\\" defaultbranchname=\\"main\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-2\\" show-inactive-tag=\\"true\\" size=\\"small\\" id=\\"test-id-2\\" name=\\"sub-repository-2\\" vcsprovider=\\"GITHUB\\" isprivate=\\"true\\" latestcommitoid=\\"test-commit-oid-2\\" defaultbranchname=\\"main\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-3\\" show-inactive-tag=\\"true\\" size=\\"small\\" id=\\"test-id-3\\" name=\\"sub-repository-3\\" vcsprovider=\\"GITHUB\\" isprivate=\\"true\\" latestcommitoid=\\"test-commit-oid-3\\" defaultbranchname=\\"main\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-4\\" show-inactive-tag=\\"true\\" size=\\"small\\" id=\\"test-id-4\\" name=\\"sub-repository-4\\" vcsprovider=\\"GITHUB\\" isprivate=\\"true\\" latestcommitoid=\\"test-commit-oid-4\\" defaultbranchname=\\"main\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-5\\" show-inactive-tag=\\"true\\" size=\\"small\\" id=\\"test-id-5\\" name=\\"sub-repository-5\\" vcsprovider=\\"GITHUB\\" isprivate=\\"true\\" latestcommitoid=\\"test-commit-oid-5\\" defaultbranchname=\\"main\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-6\\" show-inactive-tag=\\"true\\" size=\\"small\\" id=\\"test-id-6\\" name=\\"sub-repository-6\\" vcsprovider=\\"GITHUB\\" latestcommitoid=\\"test-commit-oid-6\\" defaultbranchname=\\"main\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-7\\" size=\\"small\\" id=\\"test-id-7\\" name=\\"sub-repository-7\\" vcsprovider=\\"GITHUB\\" isactivated=\\"true\\" latestcommitoid=\\"test-commit-oid-7\\" defaultbranchname=\\"master\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-8\\" size=\\"small\\" id=\\"test-id-8\\" name=\\"sub-repository-8\\" vcsprovider=\\"GITHUB\\" isactivated=\\"true\\" latestcommitoid=\\"test-commit-oid-8\\" defaultbranchname=\\"master\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-9\\" size=\\"small\\" id=\\"test-id-9\\" name=\\"sub-repository-9\\" vcsprovider=\\"GITHUB\\" isactivated=\\"true\\" latestcommitoid=\\"test-commit-oid-9\\" defaultbranchname=\\"master\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
    <repocard-stub is-exact-route=\\"true\\" owner-login=\\"deepsourcelabs\\" route=\\"sub-repository-10\\" size=\\"small\\" id=\\"test-id-10\\" name=\\"sub-repository-10\\" vcsprovider=\\"GITHUB\\" isactivated=\\"true\\" latestcommitoid=\\"test-commit-oid-10\\" defaultbranchname=\\"master\\" lastanalyzedat=\\"2022-11-20T00:00:00+05:30\\" kind=\\"SUBREPO\\"></repocard-stub>
  </div>
  <zpagination-stub page=\\"1\\" totalpages=\\"10\\" totalvisible=\\"5\\" hideforsinglepage=\\"true\\" class=\\"flex justify-center\\"></zpagination-stub>
</div>"
`;

exports[`[[ MonorepoOverview ]] renders different states based on the \`loading\` and \`showAddSubRepositoryCta\` prop combinations: {"showAddSubRepositoryCta":true,"loading":true} 1`] = `
"<div class=\\"space-y-4 max-w-3xl\\">
  <div class=\\"h-8 animate-pulse bg-ink-300\\"></div>
  <div class=\\"space-y-3\\">
    <div class=\\"h-18 animate-pulse bg-ink-300\\"></div>
    <div class=\\"h-18 animate-pulse bg-ink-300\\"></div>
    <div class=\\"h-18 animate-pulse bg-ink-300\\"></div>
    <div class=\\"h-18 animate-pulse bg-ink-300\\"></div>
  </div>
  <zpagination-stub page=\\"1\\" totalpages=\\"10\\" totalvisible=\\"5\\" hideforsinglepage=\\"true\\" class=\\"flex justify-center\\"></zpagination-stub>
</div>"
`;
